### PR TITLE
feature: initial theming

### PR DESCRIPTION
**Pull request by iamstarkov**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,4 @@ dist
 coverage
 tests
 tmp
+shrinkwrap.yaml
```
Comment (kof) [inline .gitignore]: started using npm5 lock files

File: README.md [hunk 2]
```diff
@@ -46,67 +69,112 @@ const Button = ({classes, children}) => (
 export default injectSheet(styles)(Button)
 ```
 
-## Custom setup.
+### Theming
 
-If you want to specify a jss version and plugins to use, you should create your [own jss instance](https://github.com/cssinjs/jss/blob/master/docs/js-api.md#create-an-own-jss-instance), [setup plugins](https://github.com/cssinjs/jss/blob/master/docs/setup.md#setup-with-plugins) and create a `injectSheet` function which has your jss version bound.
+Idea is easy. You define theme object once in a app and pass it to `ThemeProvider`. Later you use `withTheme` to map theme object from context to components' props. And then use `props.theme` as any other props. After that you may change your theme, and all your components will get new theme automatically.
```
Comment (kof) [inline README.md]: please don't use words like easy and simple

File: .gitignore (hunk 1 above)
Comment (iamstarkov) [inline .gitignore]: hm?

File: README.md (hunk 2 above)
Comment (iamstarkov) [inline README.md]: ok, easy is controversial. but "simple" should be okay

Comment (kof) [inline README.md]: > object once in a app and pass

mb: "object once for a subtree or the entire app"

File: README.md [hunk 3]
```diff
@@ -46,67 +69,112 @@ const Button = ({classes, children}) => (
...
+
+Useful for styleguides, so you can change brand colors and other stuff without rebuilding all your components.
```
Comment (kof) [inline README.md]: one may think its useful just for styleguides, I would  remove this sentence

Comment (iamstarkov) [inline README.md]: ok

Comment (iamstarkov) [inline README.md]: done

File: README.md [hunk 4]
```diff
@@ -46,67 +69,112 @@ const Button = ({classes, children}) => (
...
+Under the hood `react-jss` uses unified CSSinJS `theming` solution for React. You can find [detailed docs in its repo](https://github.com/iamstarkov/theming).
 
-export const injectSheet = createInjectSheet(jss)
-```
+In your app you will need `ThemeProvider`:
 
-You can also access the Jss instance being used by default.
+* it has `theme` prop which should be an `object` or `function`:
+  * If its `Object` and it is a root `ThemeProvider` then its intact and being passed down the react tree.
+  * If its `Object` and its nested `ThemeProvider` then its being merged with theme from parent `ThemeProvider` and passed down to the react tree.
```
Comment (kof) [inline README.md]: > down to the react tree.

down the react tree

Comment (iamstarkov) [inline README.md]: sure. related https://github.com/iamstarkov/theming/issues/8

File: .gitignore (hunk 1 above)
Comment (iamstarkov) [inline .gitignore]: i mean i dont get it

Comment (iamstarkov) [inline .gitignore]: but, i just remove this change

Comment (iamstarkov) [inline .gitignore]: removed

File: README.md (hunk 2 above)
Comment (iamstarkov) [inline README.md]: rephrased it to:

> Idea is simple. You define theme `object`. You pass it to `ThemeProvider` which in turn passes it to `context`, then you use `withTheme` to map it back from `context` to components `props`. After that you may change your theme, and all your components will get new theme automatically.

Comment (kof) [inline README.md]: those are just words which express a subjective opinion, for some it is simple for others it is not. There is nothing easy or simple in software engineering in general and we should not use those words. I have even seen a linter for wording real quick and those where in the blacklist I believe.  I should start using it.

Comment (iamstarkov) [inline README.md]: strongly related https://www.infoq.com/presentations/Simple-Made-Easy

Comment (iamstarkov) [inline README.md]: tldr: 

> Simple is often erroneously mistaken for easy. "Easy" means "to be at hand", "to be approachable". "Simple" is the opposite of "complex" which means "being intertwined", "being tied together". Simple != easy.

>  Easy is a relative term. Something may be easy for someone and difficult for someone else.

Comment (iamstarkov) [inline README.md]: I still think "simple" is still relevant here

File: README.md [hunk 5]
```diff
@@ -46,67 +69,110 @@ const Button = ({classes, children}) => (
...
+Idea is simple. You define theme `object`. You pass it to `ThemeProvider` which in turn passes it to `context`, then you use `withTheme` to map it back from `context` to components `props`. After that you may change your theme, and all your components will get new theme automatically.
 
-export const injectSheet = createInjectSheet(jss)
-```
+Under the hood `react-jss` uses unified CSSinJS `theming` solution for React. You can find [detailed docs in its repo](https://github.com/iamstarkov/theming).
 
-You can also access the Jss instance being used by default.
+In your app you will need `ThemeProvider`:
+
+* it has `theme` prop which should be an `object` or `function`:
```
Comment (RikardGehlin) [inline README.md]: Sentence starts with capital letters

File: README.md [hunk 6]
```diff
@@ -46,67 +69,110 @@ const Button = ({classes, children}) => (
...
+  * If its `Object` and it is a root `ThemeProvider` then its intact and being passed down the react tree.
```
Comment (RikardGehlin) [inline README.md]: It should be "it's" since it is an abbreviation of "it is".

File: README.md [hunk 7]
```diff
@@ -46,67 +69,110 @@ const Button = ({classes, children}) => (
...
+  * If its `Object` and its nested `ThemeProvider` then its being merged with theme from parent `ThemeProvider` and passed down to the react tree.
```
Comment (RikardGehlin) [inline README.md]: See above comment

File: README.md [hunk 8]
```diff
@@ -46,67 +69,110 @@ const Button = ({classes, children}) => (
...
+  * If its `Function` and its nested `ThemeProvider` then its being applied to the theme from parent `ThemeProvider`. if result is an `Object` it will be passed down the react tree, throws otherwise.
```
Comment (RikardGehlin) [inline README.md]: See above comment, and also start the second sentence with a capital letter

File: README.md [hunk 9]
```diff
@@ -46,67 +69,110 @@ const Button = ({classes, children}) => (
...
+* `ThemeProvider` as every other component can rended only single children, because it uses `React.Children.only` in render, which returns the only child in children. Throws otherwise.
```
Comment (RikardGehlin) [inline README.md]: Render*

File: README.md [hunk 10]
```diff
@@ -46,67 +69,110 @@ const Button = ({classes, children}) => (
...
+    <Button>I am a button with green background</Button>
+  </ThemeProvider>
+}
 
-## Using decorators.
+export default App;
+```
 
-You can use ES7 with [decorators](https://github.com/wycats/javascript-decorators) (using [babel-plugin-transform-decorators-legacy](https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy)).
+In your components you will need `withTheme`. Its a Higher-order Component which takes a single `React.Component` and maps a theme object from context to props. [Read more about `withTheme` in `theming`'s documentation.](https://github.com/iamstarkov/theming#withthemecomponent)
```
Comment (RikardGehlin) [inline README.md]: It's*

**Comment by iamstarkov** [comment]
@RikardGehlin thanks for feedback. it seems i fixed everything

**Comment by kof** [comment]
merged

**Comment by iamstarkov** [comment]
yay

**Comment by iamstarkov** [comment]
@kof release, maybe?

**Comment by kof** [comment]
lets finish the other PR first?

**Comment by iamstarkov** [comment]
sure

**Comment by iamstarkov** [comment]
@kof on that note. do you see any obvious flaws in my implementation?

**Pull request merged**